lxd/storage_zfs.go

```diff
@@ -588,14 +672,64 @@ func (s *storageZfs) StoragePoolVolumeUpdate(writable *api.StorageVolumePut, cha
 	}
 
 	if shared.StringInSlice("size", changedConfig) {
-		// apply quota
 		if s.volume.Config["size"] != writable.Config["size"] {
-			size, err := shared.ParseByteSizeString(writable.Config["size"])
-			if err != nil {
-				return err
+			var err error
+			var data interface{}
+			var volumeType int
+
+			poolName := s.getOnDiskPoolName()
+			size := int64(0)
+			if writable.Config["size"] != "" {
+				size, err = shared.ParseByteSizeString(writable.Config["size"])
+				if err != nil {
+					return err
+				}
 			}
+			switch s.volume.Type {
+			case storagePoolVolumeTypeNameContainer:
+				ct, err := containerLoadByName(s.s, s.volume.Name)
+				if err != nil {
+					return err
+				}
 
-			err = s.StorageEntitySetQuota(storagePoolVolumeTypeCustom, size, nil)
+				// Quotas do not apply to snapshots.
+				if ct.IsSnapshot() {
+					return fmt.Errorf(`Quota do not apply ` +
+						`to storage volumes of snapshots`)
+				}
+
+				// Non-storage volume "size" properties override
+				// storage volume "size" properties
+				expandedDevices := ct.ExpandedDevices()
+				_, rootDiskDevice, _ := containerGetRootDiskDevice(expandedDevices)
+				if rootDiskDevice["size"] != "" {
+					return fmt.Errorf(`The storage volume ` +
+						`for this container has a ` +
+						`"size" property set on its ` +
+						`root disk device`)
+				}
```

Comment:
So this is going to get confusing as if you already have "size" set on the disk entry, then you get the error.
But if you don't, then you can set the size on the volume, then go and set the size on the disk entry too and won't get an error until you go to change any other setting on the volume.

So I think what we need in general is:
 - custom volumes => Quota (size property) is only allowed on the volume itself
 - image volumes => The size property is read-only, its value is dependent on backend.
    - dir => unset
    - btrfs => unset
    - zfs => dataset size after the image got unpacked
    - lvm => size of the filesystem
    - ceph => size of the filesystem
 - container volumes => Quota (size property) is only allowed on the container disk entry and is shown as a read-only value on the volume.

We just need to make sure that those cases have proper error handling:
 - Attempting to set the "size" property on a disk entry for a custom volume => Say that custom volume sizes must be changed at the volume level
 - Attempting to set the "size" property on an image volume => Say that images are read-only and can't be resized.
 - Attempting to set the "size" property on a container volume => Say that container volume sizes must be changed at the container disk level

I realize how widely inconsistent this all is, but I think it's the only way we can make this all work considering all the edge cases (profiles and device inheritance mostly). The main reasons being:
 - It makes sense for users to want to specify all their limits in a profile, including disk limits for the container
 - Custom volumes can't have their size set as part of the disk entry since they can be shared between multiple containers
 - We can't sync the "size" volume property and the "size" container disk property as there may not be such a specific disk property (if the disk was inherited through profiles)

I think we need to come up with a clear policy (like the one I proposed above), have it documented in storage.md and then implemented consistently across all storage backends.

What do you think?